### PR TITLE
feat: meilisearch backend

### PR DESCRIPTION
**Pull request by qasimgulzar**
Under this PR I have implemented a search backend for meilisearch.

Below is a screenshot of "discover now" page it is working with meilisearch backend.

image

Thread : https://discuss.openedx.org/t/is-meilisearch-a-viable-upgrade-alternative-to-opensearch/12400/12

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @qasimgulzar! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: search/meilisearch.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+            source["usage_key"] = source["id"]
+        source["id"] = usage_key.block_id
+    except (Exception, InvalidKeyError) as ex:  # pylint: disable=broad-except
+        source["id"] = sanitize_id(source["id"])
+        log.info(f"{str(ex)} - {source['id']} - {type(ex)}")
+
+    return source
+
+
+def build_filter(key, val):
```
Comment (bradenmacdonald) [inline search/meilisearch.py:91]: All of these functions will need extensive tests cases.

File: search/meilisearch.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    def set_mappings(cls, index_name, mappings):
+        """
+        Set new mapped-items structure into cache
+        """
+        cache.set(cls.get_cache_item_name(index_name), mappings)
+
+    @property
+    def mappings(self):
+        """
+        Get mapping of current index.
```
Comment (bradenmacdonald) [inline search/meilisearch.py:268]: What are the "mappings" ?

File: search/meilisearch.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    def _clear_mapping(self):
+        """
+        Remove the cached mappings.
+        """
+        MeiliSearchEngine.set_mappings(self._prefixed_index_name, {})
+
+    def __init__(self, index=None, options=None):
+        super().__init__(index)
+        MEILISEARCH_URL = getattr(settings, "MEILISEARCH_URL", 'http://127.0.0.1:7700')
+        MEILISEARCH_API_KEY = getattr(settings, "MEILISEARCH_API_KEY", "masterKey")
```
Comment (bradenmacdonald) [inline search/meilisearch.py:289]: I don't know if we should be putting the default values here - we might get inconsistent defaults if we just define the defaults separately in each place where we read the setting.

**Review by bradenmacdonald — COMMENTED** [review]
Does this PR make courseware search work using Meilisearch or just add a new "auto-suggest" search that works using Meilisearch? Or both?

File: search/meilisearch.py [hunk 4]
```diff
@@ -0,0 +1,389 @@
+"""
+Meilisearch implementation for courseware search index
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply search/meilisearch.py:2]: this is not only for courseware search, we can use this backend to other endpoints as well.

File: edxsearch/settings.py [hunk 5]
```diff
@@ -26,7 +26,7 @@
 # This is just a container for running tests
 DEBUG = True
 
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = ['*']
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply edxsearch/settings.py:29]: It is mostly used during development

File: search/api.py [hunk 6]
```diff
@@ -158,3 +160,102 @@ def course_discovery_search(search_term=None, size=20, from_=0, field_dictionary
     )
 
     return results
+
+
+def _meilisearch_auto_suggest_search_api(term, course_id, limit=30):
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply search/api.py:165]: removing this endpoint

File: search/api.py [hunk 7]
```diff
@@ -158,3 +160,102 @@ def course_discovery_search(search_term=None, size=20, from_=0, field_dictionary
...
+    """
+    Perform an auto-suggest search using the Elasticsearch search engine.
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply search/api.py:167]: It was a POC for implementation I am going to remove it

File: search/meilisearch.py [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+from opaque_keys.edx.keys import UsageKey
+from search.search_engine_base import SearchEngine
+from search.utils import ValueRange, _is_iterable
+
+# log appears to be standard name used for logger
+log = logging.getLogger(__name__)
+
+prefix = getattr(settings, "MEILISEARCH_INDEX_PREFIX", "")
+RESERVED_CHARACTERS = "+=><!(){}[]^~*:\\/&|?"
+INDEX_SETTINGS = {
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply search/meilisearch.py:22]: I can probably move this to settings

File: search/meilisearch.py [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+
+def sanitize_id(_id: str | int) -> str:
+    return hashlib.md5(f"{_id}".encode('utf-8')).hexdigest()
+
+
+def sanitized_id(source: dict, create_usage_key=True) -> dict:
+    """
+    Sanitize the Id key to avoid restricted objects
```
[the remark this answers is not in the document]
Comment (qasimgulzar) [reply search/meilisearch.py:71]: Meilisearch doesn't allow to put course-key in ID attribute so to make it compatible I implemented this function.

**Comment by e0d** [comment]
Your branch is behind the base.  I've pulled in changes from master as a merge commit which will update your branch and cause the tests to be re-run. 

@qasimgulzar should you be listed under the Arbisoft entity CLA?  Currently we have to run tests on your PR manually.

**Comment by regisb** [comment]
Yes, I confirm that @qasimgulzar is part of Arbisoft and should be added to the company CLA. Is there any step that we should pursue on our side?

**Comment by mphilbrick211** [comment]
Hi @qasimgulzar! I see your CLA is all set. Do you plan to pursue this pull request?

**Comment by qasimgulzar** [comment]
We am working on another extension but we might need this for further testing.

**Comment by qasimgulzar** [comment]
Hi @e0d, I hope you are doing great. I will no longer be part of edly from 7, OCT 2024. could you please keep my CLA back to individual and remove me from arbisoft's CLA.

**Comment by mphilbrick211** [comment]
Hi @qasimgulzar! Can this pull request be closed?

**Comment by qasimgulzar** [comment]
closing this PR, as it is not intended to merged.

**Pull request closed without merging**